### PR TITLE
Handles User Object Validation #142

### PR DESCRIPTION
**Pull request by freddyC**
Added in validation for training options and tests as well.

A GIF or MEME to give some spice of the internet

## Description

Creates validation for when the user sets the training options. 

## Motivation and Context

[issue](https://github.com/BrainJS/brain.js/issues/142)
This should help users who run into problems to see things they may have done wrong.

## How Has This Been Tested?

I created a slew of tests that go over the expected supported functionality

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Author's Checklist:

- [x] My code focuses on the main motivation and avoids scope creep.
- [x] My code passes current tests and adds new tests where possible.
- [x] My code is [SOLID](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)) and [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).
- [x] I have updated the documentation as needed.

## Reviewer's Checklist:
- [ ] I kept my comments to the author positive, specific, and productive.
- [ ] I tested the code and didn't find any new problems.
- [ ] I think the motivation is good for the project.
- [ ] I think the code works to satisfies the motivation.

### CODE REVIEW

File: src/neural-network.js [hunk 1]
```diff
@@ -35,12 +35,27 @@ export default class NeuralNetwork {
     };
   }
 
+  static get trainingValidations() {
```
Comment (perkyguy) [inline src/neural-network.js]: Do we need a `get` for this? Why not just define these in the validation method itself?

File: src/neural-network.js [hunk 2]
```diff
@@ -35,12 +35,27 @@ export default class NeuralNetwork {
...
+    return {
+      iterations: (val) => { return typeof val === 'number' && val > 0; },
+      errorThresh: (val) => { return typeof val === 'number' && val > 0 && val < 1; },
+      log: (val) => { return typeof val === 'function' || typeof val === 'boolean'; },
+      logPeriod: (val) => { return typeof val === 'number' && val > 0; },
```
Comment (perkyguy) [inline src/neural-network.js]: logPeriod only matters if you also defined `log` as well. I think these functions don't _have_ to be a one-to-one mapping for the training options and can instead explicitly check the training options they are testing. This way, they all would be of the form `(opts) => { ... }`.

File: src/neural-network.js [hunk 3]
```diff
@@ -35,12 +35,27 @@ export default class NeuralNetwork {
...
+    }
+  }
+
   constructor(options = {}) {
     Object.assign(this, this.constructor.defaults, options);
     this.hiddenSizes = options.hiddenLayers;
     this.trainOpts = {};
     this._updateTrainingOptions(Object.assign({}, this.constructor.trainDefaults, options));
 
+    this._isStrict = true;
```
Comment (perkyguy) [inline src/neural-network.js]: If we are going to have multiple validation options, this should be contained within an object that contains the options and this as well. If we are not going to have multiple validation options, this should be named to something more descriptive, e.g. `this._strictTrainingOptionValidation` or something

File: src/neural-network.js [hunk 4]
```diff
@@ -60,6 +75,16 @@ export default class NeuralNetwork {
     }
   }
 
+  /**
+   * 
+   * validateTrainingOptions is set true as default
+   *      When true: setting a training option outside expected range will throw an error
+   *      When false: setting a training option outside expected range will log out and continue
+   */
+  get validateTrainingOptions() { return this._isStrict; }
```
Comment (perkyguy) [inline src/neural-network.js]: Return an object? the name suggests there are multiple

File: src/neural-network.js [hunk 5]
```diff
@@ -60,6 +75,16 @@ export default class NeuralNetwork {
...
+  set validateTrainingOptions(s) { this._isStrict = s; }
```
Comment (perkyguy) [inline src/neural-network.js]: pass in an object?

File: src/neural-network.js [hunk 6]
```diff
@@ -366,6 +392,22 @@ export default class NeuralNetwork {
     return true;
   }
 
+
+  /**
+   *
+   * @param options
+   * @private
+   */
+  _validateTrainingOptions(options) {
```
Comment (perkyguy) [inline src/neural-network.js]: This can be `static`, don't think it matters, but it can be.

File: src/neural-network.js [hunk 7]
```diff
@@ -366,6 +392,22 @@ export default class NeuralNetwork {
...
+    Object.keys(options).forEach(key => {
+      if (NeuralNetwork.trainingValidations[key] && !NeuralNetwork.trainingValidations[key](options[key])) {
+        let message = `[${key}, ${options[key]}] is out of noraml range`;
+        if (this._isStrict) throw new Error (message);
+        else console.log(message);
```
Comment (perkyguy) [inline src/neural-network.js]: `console.error`? `console.warn`?

File: src/neural-network.js [hunk 8]
```diff
@@ -366,6 +392,22 @@ export default class NeuralNetwork {
...
+  /**
+   *
+   * @param options
+   * @private
+   */
+  _validateTrainingOptions(options) {
+    Object.keys(options).forEach(key => {
+      if (NeuralNetwork.trainingValidations[key] && !NeuralNetwork.trainingValidations[key](options[key])) {
+        let message = `[${key}, ${options[key]}] is out of noraml range`;
+        if (this._isStrict) throw new Error (message);
```
Comment (perkyguy) [inline src/neural-network.js]: If these settings are wrong, this will break training, right? Why not just always enforce `isStrict`?

File: src/neural-network.js [hunk 9]
```diff
@@ -366,6 +392,22 @@ export default class NeuralNetwork {
...
+
+  /**
+   *
+   * @param options
+   * @private
+   */
+  _validateTrainingOptions(options) {
+    Object.keys(options).forEach(key => {
+      if (NeuralNetwork.trainingValidations[key] && !NeuralNetwork.trainingValidations[key](options[key])) {
+        let message = `[${key}, ${options[key]}] is out of noraml range`;
```
Comment (perkyguy) [inline src/neural-network.js]: typo in "normal"

File: src/neural-network.js [hunk 10]
```diff
@@ -366,6 +392,22 @@ export default class NeuralNetwork {
...
   }
 
+
+  /**
+   *
+   * @param options
+   * @private
+   */
+  _validateTrainingOptions(options) {
+    Object.keys(options).forEach(key => {
```
Comment (perkyguy) [inline src/neural-network.js]: Why not iterate over the `trainingValidations` functions? Only care about things we care about?

File: test/base/bitwise.js [hunk 11]
```diff
@@ -75,30 +75,30 @@ describe('bitwise functions async training', () => {
     let not = [{input: [0], output: [1]},
                {input: [1], output: [0]}];
     testBitwiseAsync(not, 'not', done);
-  }).timeout(10000);
+  }).timeout(15000);
```
Comment (perkyguy) [inline test/base/bitwise.js]: why the change?

File: test/base/bitwise.js [hunk 12]
```diff
@@ -75,30 +75,30 @@ describe('bitwise functions async training', () => {
...
 
   it('XOR function', (done) => {
     let xor = [{input: [0, 0], output: [0]},
                {input: [0, 1], output: [1]},
                {input: [1, 0], output: [1]},
                {input: [1, 1], output: [0]}];
     testBitwiseAsync(xor, 'xor', done);
-  }).timeout(10000);
+  }).timeout(15000);
```
Comment (perkyguy) [inline test/base/bitwise.js]: Why the change?

File: test/base/bitwise.js [hunk 13]
```diff
@@ -75,30 +75,30 @@ describe('bitwise functions async training', () => {
...
 
   it('OR function', (done) => {
     let or = [{input: [0, 0], output: [0]},
               {input: [0, 1], output: [1]},
               {input: [1, 0], output: [1]},
               {input: [1, 1], output: [1]}];
     testBitwiseAsync(or, 'or', done);
-  }).timeout(10000);
+  }).timeout(15000);
```
Comment (perkyguy) [inline test/base/bitwise.js]: Why the change?

File: test/base/bitwise.js [hunk 14]
```diff
@@ -75,30 +75,30 @@ describe('bitwise functions async training', () => {
...
 
   it('AND function', (done) => {
     let and = [{input: [0, 0], output: [0]},
                {input: [0, 1], output: [0]},
                {input: [1, 0], output: [0]},
                {input: [1, 1], output: [1]}];
     testBitwiseAsync(and, 'and', done);
-  }).timeout(10000);
+  }).timeout(15000);
```
Comment (perkyguy) [inline test/base/bitwise.js]: Why the change?

File: test/base/trainopts.js [hunk 15]
```diff
@@ -173,3 +173,98 @@ describe('async train() options', () => {
...
+    assert.throws(() => { net._updateTrainingOptions({ errorThresh: false}) });
+    assert.doesNotThrow(() => { net._updateTrainingOptions({ errorThresh: 0.008}) });
+  });
+
+  it('log validation', () => {
+    let net = new brain.NeuralNetwork();
+    assert.throws(() => { net._updateTrainingOptions({ log: 'no strings' }) });
+    assert.throws(() => { net._updateTrainingOptions({ log: 4 }) });
+    assert.doesNotThrow(() => { net._updateTrainingOptions({ log: false }) });
+    assert.doesNotThrow(() => { net._updateTrainingOptions({ log: () => {} }) });
```
Comment (perkyguy) [inline test/base/trainopts.js:196]: is this a "valid" log function?

File: test/base/trainopts.js [hunk 16]
```diff
@@ -173,3 +173,98 @@ describe('async train() options', () => {
...
+    assert.doesNotThrow(() => { net._updateTrainingOptions({ momentum: 0.8 }) });
+  });
+
+  it('callback validation', () => {
+    let net = new brain.NeuralNetwork();
+    assert.throws(() => { net._updateTrainingOptions({ callback: 'no strings' }) });
+    assert.throws(() => { net._updateTrainingOptions({ callback: 4 }) });
+    assert.throws(() => { net._updateTrainingOptions({ callback: false }) });
+    assert.doesNotThrow(() => { net._updateTrainingOptions({ callback: null }) });
+    assert.doesNotThrow(() => { net._updateTrainingOptions({ callback: () => {} }) });
```
Comment (perkyguy) [inline test/base/trainopts.js:234]: Is this a "valid" callback?

**Review by perkyguy — CHANGES_REQUESTED** [review]

File: test/base/bitwise.js (hunk 11 above)
Comment (freddyC) [reply test/base/bitwise.js]: I had a test that timed out the first time.  I am actually thinking we should get rid of the async tests as much as possible (now that the training ticks are shared)  They really delay the test suit

File: test/base/trainopts.js (hunk 15 above)
Comment (freddyC) [reply test/base/trainopts.js:196]: yep, it would be user defined.  (logging allows the user to either set true and it will console.log or set their own method

File: test/base/trainopts.js (hunk 16 above)
Comment (freddyC) [reply test/base/trainopts.js:234]: yep (it just does nothing with the passed in object, I am not enforcing the params)

File: src/neural-network.js (hunk 1 above)
Comment (freddyC) [reply src/neural-network.js]: sorry I didn't see your comments but yeah I got rid of the getter/setter.  Originally when they set this to false I was printing out a message showing what it would do, but I thought that overkill.

File: test/base/bitwise.js (hunk 11 above)
Comment (perkyguy) [reply test/base/bitwise.js]: definitely should change up how we test the async. however, that should be done in a separate PR and I'd say just revert this back to limit unnecessary changes of this PR.

File: test/base/trainopts.js (hunk 15 above)
Comment (perkyguy) [reply test/base/trainopts.js:196]: I don't understand why the `log` value can be user defined, that's the point of `callback`. Buuuuuut, I was a bit too hasty with this and the other "Valid callback" comment. This is fine with me.

File: src/neural-network.js (hunk 2 above)
Comment (freddyC) [reply src/neural-network.js]: I am not sure I completely understand your point.  I could couple them so that we only check things that are dependent on each other and I could set them to check the value of the already set property.  I chose not to because I wanted to keep each property decoupled.  I think that will make it easier to extend (if you change the function of one property you don't have to worry about additional side effects) and I also didn't want to check every property each time I validate, just the properties that are changed.  Does that make sense?  I am open to changing but I don't yet understand the benefit.

File: src/neural-network.js (hunk 3 above)
Comment (freddyC) [reply src/neural-network.js]: totally agree the name didn't fit.  I ended up changing it before I saw this comment.

File: src/neural-network.js (hunk 4 above)
Comment (freddyC) [reply src/neural-network.js]: I don't see how that name suggests an object is returned.  I did change it from being a getter/setter. Originally I was printing out the a log saying what was happening when you set it to false but I thought that was a bit overkill especially once someone understands already how it works.

File: src/neural-network.js (hunk 5 above)
Comment (freddyC) [reply src/neural-network.js]: yeah it was a setter/getter but I got rid of that (since there aren't any intentional side effects)

File: src/neural-network.js (hunk 6 above)
Comment (freddyC) [reply src/neural-network.js]: Good idea, there really is no need to have it on the object instead of on the class.

File: src/neural-network.js (hunk 10 above)
Comment (freddyC) [reply src/neural-network.js]: yeah I only wanted to check the things that actually are changing.  I could see it going either way (the cost isn't huge).

File: src/neural-network.js (hunk 8 above)
Comment (freddyC) [reply src/neural-network.js]: I thought it would let the class stay more flexible to allow the validation to be shut off (in the odd case there is some weird option that they legitimately want to allow).  I kinda wrestled with this, it's that balance between providing structure and allowing innovation/creativity.  I landed on this due to the goal of the ticket, to help those who are new figure out things that might be going wrong.  Enforcing unexpected behavior I felt didn't fall within that (even though I can't think of a reason why they would want to do that).

File: src/neural-network.js (hunk 7 above)
Comment (freddyC) [reply src/neural-network.js]: Good idea, I'll change it to console.warn

File: test/base/bitwise.js (hunk 11 above)
Comment (freddyC) [reply test/base/bitwise.js]: 👍

File: test/base/trainopts.js (hunk 15 above)
Comment (freddyC) [reply test/base/trainopts.js:196]: No worries.  The main idea is so that if the user has their own logging system (i.e. it writes to file or whatever) then they can keep the log to do just that, without having to use up their callback for that.

Comment (perkyguy) [reply test/base/trainopts.js:196]: Meh, I don't think it would "use up" their callback. But this is a moot point, we can continue discussion outside of this PR.

**Comment by freddyC** [comment]
@perkyguy I think this is ready for another look-see

File: src/neural-network.js [hunk 17]
```diff
@@ -35,12 +35,40 @@ export default class NeuralNetwork {
...
+    });
+  }
+
   constructor(options = {}) {
     Object.assign(this, this.constructor.defaults, options);
     this.hiddenSizes = options.hiddenLayers;
     this.trainOpts = {};
     this._updateTrainingOptions(Object.assign({}, this.constructor.trainDefaults, options));
 
+    this.invalidTrainOptsShouldThrow = true;
```
Comment (perkyguy) [inline src/neural-network.js]: why put this on the instance and not just always require valid options? Brain won't train correctly without these set?

File: src/neural-network.js [hunk 18]
```diff
@@ -35,12 +35,40 @@ export default class NeuralNetwork {
...
+      errorThresh: (val) => { return typeof val === 'number' && val > 0 && val < 1; },
+      log: (val) => { return typeof val === 'function' || typeof val === 'boolean'; },
+      logPeriod: (val) => { return typeof val === 'number' && val > 0; },
+      learningRate: (val) => { return typeof val === 'number' && val > 0 && val < 1; },
+      momentum: (val) => { return typeof val === 'number' && val > 0 && val < 1; },
+      callback: (val) => { return typeof val === 'function' || val === null },
+      callbackPeriod: (val) => { return typeof val === 'number' && val > 0; },
+      timeout: (val) => { return typeof val === 'number' && val > 0 }
+    };
+    Object.keys(options).forEach(key => {
```
Comment (perkyguy) [inline src/neural-network.js]: Only reason I would go over all the validation keys rather than the passed in objects, was to ensure validity of all options?

**Review by perkyguy — COMMENTED** [review]
Questions more than anything

**Comment by robertleeplummerjr** [comment]
@perkyguy you waiting for anything else here?

**Review by perkyguy — APPROVED** [review]

**Comment by perkyguy** [comment]
I think I'm good with all of this. Fix the merge conflicts though.

**Comment by perkyguy** [comment]
I think I'm good with all of this. Fix the merge conflicts though, and make sure you build prior to merging.

**Comment by robertleeplummerjr** [comment]
nice
